src/ert/ensemble_evaluator/dispatch.py

```diff
@@ -74,20 +60,47 @@ def done_logger(_):
         events_handling.add_done_callback(done_logger)
         await events_handling
 
+    def run_dispatcher(self, loop):
+        try:
+            loop.run_until_complete(self._job())
+        except asyncio.CancelledError as ex:
+            logger.warning(f"batcher was cancelled: {ex}")
+        except Exception as failure:  # pylint: disable=broad-exception-caught
+            logger.exception(f"exception in batching dispatcher: {failure}")
+            trace_info = traceback.format_exception(
+                type(failure), failure, failure.__traceback__
+            )
+            logger.error(f"{trace_info}")
+        else:
+            logger.debug("batcher finished normally")
+            return
+        finally:
+            self._finished = True
+        loop.run_until_complete(self._done_callback())
+        logger.debug("Dispatcher thread exiting.")
+
+    async def _done_callback(self):
+        # call any registered handlers for FAILED. since we don't have
+        # an event, pass empty list and let handler decide how to proceed
+        funcs = self._LOOKUP_MAP[identifiers.EVTYPE_ENSEMBLE_FAILED]
+        await asyncio.gather(*[f([]) for f, _ in funcs])
+
     async def _job(self):
         while self._running:
-            await asyncio.sleep(self._timeout)
+            if len(self._buffer) < self._max_batch:
+                time.sleep(self._timeout)
+            else:
+                time.sleep(0)
             await self._work()
-
         # Make sure no events are lingering
         await self._work()
 
     async def join(self):
         self._running = False
         # if result is exception it should have been handled by
         # done-handler, but also avoid killing the caller here
-        with contextlib.suppress(BaseException):
-            await self._task
+        while not self._finished:
+            await asyncio.sleep(0.01)
 
     def register_event_handler(self, event_types, function, batching=True):
```

Comment:
This seems to be always true (by default we always use batching). Maybe removing the parameter would simplify the code (removing the `if batching`) as I don't see that we can get rid of it any time soon.

Reply:
removed the logic and renamed `timeout` - see if you like it:
https://github.com/equinor/ert/pull/5828/commits/31ba8baff50d190b1037540a658184b018ac8836

Reply:
Very good! :smile: